File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    LinkedList<Integer> path = new LinkedList<>();
    List<List<Integer>> ans = new ArrayList<>();

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        Arrays.sort(candidates);
        backTracking(candidates, target, 0, 0);
        return ans;
    }

    private void backTracking(int[] candidates, int targetSum, int sum, int startIdx) {
        if (sum == targetSum) {
            ans.add(new ArrayList<>(path));
        }

        for (int i = startIdx; i < candidates.length; i++) {
            if (sum > targetSum) {
                break;
            }
            path.add(candidates[i]);
            sum += candidates[i];
            backTracking(candidates, targetSum, sum, i);
            sum -= candidates[i];
            path.removeLast();
        }
    }
}

File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        ans = []

        def backtracking(startIdx: int, path: List[int], curr_sum: int) -> None:
            if curr_sum == target:
                ans.append(path[:])
                return
            
            if curr_sum > target:
                return
            
            for i in range(startIdx, len(candidates)):
                curr_sum += candidates[i]
                path.append(candidates[i])
                backtracking(i, path, curr_sum)
                path.pop()
                curr_sum -= candidates[i]
        
        backtracking(0, [], 0)
        return ans